akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala

```diff
@@ -96,6 +102,7 @@ private[http] object HttpServerBluePrint {
 
     override def createLogic(inheritedAttributes: Attributes) = new GraphStageLogic(shape) {
```

Comment:
In order to save even more allocations this logic could extend `InHandler with OutHandler`, putting the idle case on it directly.